File: rust/src/codec/slot.rs
//! Slot codec — inventory item-stack data.

use crate::codec::{Reader, Writer, nbt, varint};
use crate::errors::ProtocolError;

/// Populated inventory slot.
#[derive(Debug, Clone, PartialEq)]
pub struct SlotData {
    /// Numeric item registry id.
    pub item_id: i32,
    /// Stack count (i8 range).
    pub count: i8,
    /// Optional NBT tag.
    pub tag: Option<nbt::NbtTag>,
}

/// Decode a slot. Returns `None` when the slot is empty (`present = 0`).
pub fn read<R: Reader + ?Sized>(reader: &mut R) -> Result<Option<SlotData>, ProtocolError> {
    let present = reader.read_exact(1)?[0];
    match present {
        0 => Ok(None),
        1 => {
            let item_id = varint::read(reader)?;
            let count = reader.read_exact(1)?[0] as i8;
            let tag = nbt::read(reader)?;
            Ok(Some(SlotData { item_id, count, tag }))
        }
        other => Err(ProtocolError::EncodeError(format!(
            "slot.present: {other} (expected 0 or 1)"
        ))),
    }
}

/// Encode an `Option<SlotData>` as a Slot.
pub fn write<W: Writer + ?Sized>(
    value: Option<&SlotData>,
    writer: &mut W,
) -> Result<(), ProtocolError> {
    match value {
        None => writer.write_all(&[0]),
        Some(s) => {
            writer.write_all(&[1])?;
            varint::write(s.item_id, writer)?;
            writer.write_all(&(s.count as u8).to_be_bytes())?;
            nbt::write(s.tag.as_ref(), writer)
        }
    }
}

File: python/minecraft_bot/codec/slot.py
"""Slot codec — inventory item-stack data.

Wire format for protocol 763 (1.20.1)::

    bool present
    if present:
        VarInt item_id
        i8    count
        NBT   tag           # may be a single TAG_End byte for "no NBT"

Empty slot is encoded as a single ``0x00`` byte (``present = false``).

The Python side returns / accepts ``Optional[SlotData]`` — ``None``
means the slot is empty.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, nbt, varint
from minecraft_bot.errors import ValueOutOfRange


@dataclass(frozen=True, slots=True)
class SlotData:
    """A populated inventory slot."""

    item_id: int           # numeric item registry ID for the protocol version
    count: int             # i8; protocol allows -1..127, but typical range is 0..64
    tag: nbt.NbtTag | None = None  # item NBT, e.g. enchantments, custom name


def read(reader: Reader) -> SlotData | None:
    """Decode a slot. Returns ``None`` if the slot is empty (present=false)."""
    present = reader.read(1)[0]
    if present == 0:
        return None
    if present != 1:
        # The protocol only specifies 0 / 1; treat anything else as malformed.
        raise ValueOutOfRange("slot.present", present)
    item_id = varint.read(reader)
    (count,) = struct.unpack(">b", reader.read(1))
    tag = nbt.read(reader)
    return SlotData(item_id=item_id, count=count, tag=tag)


def write(value: SlotData | None, writer: Writer) -> None:
    """Encode a slot. ``None`` writes a single ``0x00`` byte."""
    if value is None:
        writer.write(b"\x00")
        return
    if not -128 <= value.count <= 127:
        raise ValueOutOfRange("slot.count", value.count)
    writer.write(b"\x01")
    varint.write(value.item_id, writer)
    writer.write(struct.pack(">b", value.count))
    nbt.write(value.tag, writer)


__all__ = ["SlotData", "read", "write"]
